wrangler-api/src/main/java/io/cdap/wrangler/api/Directive.java

```diff
@@ -126,4 +132,44 @@ default List<EntityCountMetric> getCountMetrics() {
     // no op
     return null;
   }
+
+  /**
+   * This method provides a way to specify the output {@link Schema} generated by a directive after it performs its
+   * transformation on the data.
+   * @param inputSchema schema of the input data
+   * @param generator {@link SchemaFieldGenerator} which can generate a {@link Schema.Field}
+   *                                              given a field name and value (Java object)
+   * @param output output data generated by directive after its execution
+   * @return {@link Schema} of the output data
+   * @throws SchemaGenerationException when there is a problem with generating schema for a Java object
+   * @implNote By default, the implementation has the following steps:
+   * <ol>
+   *   <li>Get map of renamed fields and map of type changed fields</li>
+   *   <li>Iterate over list of fields in each output row</li>
+   *   <li>If output field is renamed, get original name.</li>
+   *   <li>Check if it is in map of type changed fields or if it exists in inputSchema, then use that schema</li>
+   *   <li>If it is a new field, generate the schema using provided {@link SchemaFieldGenerator}</li>
+   *   <li>Finally, add the created field to set of output fields</li>
+   * </ol>
+   */
+  @Override
+  default Schema getOutputSchema(Schema inputSchema, SchemaFieldGenerator generator, List<Row> output)
```

Comment:
I would separate responsibilities. The directive should be able to provide schema if it knows how to do it, otherwise return null. And caller should handle null and do detection

Reply:
Done, moved the schema generation handling logic to RecipePipelineExecutor